server/auth/auth.go

```diff
@@ -122,35 +123,49 @@ func (a *Auth) Authenticate(ctx context.Context, username, password, otp string,
 }
 
 func (a *Auth) authenticate(ctx context.Context, authenticator Authenticator, username, password, otp string) (context.Context, error) {
-	user, err := authenticator.Authenticate(username, password, otp)
+	session, err := authenticator.Authenticate(username, password, otp)
 	if err != nil {
 		return ctx, err
 	}
 
-	ctx = WithUser(ctx, user)
+	ctx = WithSession(ctx, session)
 
 	if a.Authorizer != nil {
-		if err := a.Authorizer.Authorize(user); err != nil {
+		if err := a.Authorizer.Authorize(session.User); err != nil {
 			return ctx, err
 		}
 	}
 
 	return ctx, nil
 }
 
+// Session represents an authenticated Session.
+type Session struct {
+	// The authenticated User.
+	User *empire.User
+
+	// When this Session will expire. The zero value means no expiration.
+	ExpiresAt *time.Time
```

Comment:
This allows us to propagate the `SessionNotOnOrAfter` assertion down, and ensures that access tokens created expire with the SAML session.